### PR TITLE
Smart import

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**
The purpose of this pull request is to allow @fbricon to review progress that I've made on the smart import feature for BuildShip. This pull request is meant as a means of communication, and is not actually meant to be pulled into the master branch of my fork.

### CODE REVIEW

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java [hunk 1]
```diff
@@ -59,131 +59,135 @@
  */
 public final class ProjectImportJob extends ToolingApiWorkspaceJob {
 
-    private final FixedRequestAttributes fixedAttributes;
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java]: I've left these changes in as a reminder to myself, and to prompt a short conversation on style. I have a habit of using the `CTRL+A CTRL+I` commands in Eclipse to reindent my code, but the indentation is not the same as the indentation that is used by the other BuildShip developers. This is easy to get around here (just don't do those commands), however, files that I create will be indented like this, and then indented differently than other Buildship files. Any suggestions on what I can do to easily maintain the indentation formatting that Buildship is currently using?

File: org.eclipse.buildship.ui.importer/about.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+<html xmlns="http://www.w3.org/1999/xhtml">
+<head>
+<meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"/>
+<title>About</title>
+</head>
+<body lang="EN-US">
+<h2>About This Content</h2>
+ 
+<p>June 5, 2006</p>	
+<h3>License</h3>
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/about.html]: @fbricon I was initially instructed to copy the `org.eclipse.buildship.ui` project and use it as a base for building the `org.eclipse.buildship.ui.importer` project. I'm uncertain which files I should keep, and which ones I should throw away. The `ui` project came with *many* .png files, which I have discarded. If you see any files that shouldn't be in this pull request, please let me know.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/wizard/project/ProjectImportWizardController.java [hunk 3]
```diff
@@ -184,7 +184,7 @@ public Job performPreviewProject(FutureCallback<Pair<OmniBuildEnvironment, OmniG
     }
 
     public boolean performImportProject() {
-        ProjectImportJob importJob = new ProjectImportJob(this.configuration);
+        ProjectImportJob importJob = new ProjectImportJob(null, this.configuration);
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/wizard/project/ProjectImportWizardController.java]: This is probably one of the worst side effects of this pull request. This has been my solution so far, as I need the `OmniEclipseProject` to mimic the functionality of the `importProject` method from the `ProjectImportJob` class. Ultimately, I don't think that this is a good route to go, but I have so far been unable to find a different path. Nevertheless, I will keep looking. Any and all suggestions and complaints appreciated.

File: tooling-e45.target [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<?pde version="3.8"?><target name="gradleware-tooling-mars" sequenceNumber="46">
+<?pde version="3.8"?><target name="gradleware-tooling-mars" sequenceNumber="47">
```
Comment (Ian-Stewart-Binks) [inline tooling-e45.target]: Uncertain why the sequenceNumber increased: Why does this happen, what does this mean?

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+		expect:
+		project.hasNature(GradleProjectNature.ID) == addGradleBuildFile
+
+		where:
+		projectName << ["GradleProject", "OtherProject"]
+		addGradleBuildFile << [true, false]
+	}
+
+	def "An existing Gradle prefs file is not overwritten"() {
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy]: Since this implementation uses the `ProjectImportJob`, this method actually *fails*. `ProjectImportJob` is responsible for this test failing, as it is responsible for creating the `.settings/gradle.prefs` file.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	@Override
+	public boolean canConfigure(IProject project, Set<IPath> ignoredPaths, IProgressMonitor monitor) {
+		return shouldBeAnEclipseProject(project, monitor);
+	}
+
+	@Override
+	public void configure(IProject project, Set<IPath> excludedDirectories, IProgressMonitor monitor) {
+		if (!project.getFile("build.gradle").exists()) return;
+
+		addNature(project, GradleProjectNature.ID, monitor);
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: One of the difficulties is that the current importing implementation uses `DefaultWorkspaceOperations` to create a project, which includes adding natures (via the `createProject` method). The `IProject` is packaged by Easymport, so we don't have the ability to create another project the same way.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+public class GradleProjectConfigurator implements ProjectConfigurator {
+
+	@Override
+	public boolean canConfigure(IProject project, Set<IPath> ignoredPaths, IProgressMonitor monitor) {
+		return shouldBeAnEclipseProject(project, monitor);
+	}
+
+	@Override
+	public void configure(IProject project, Set<IPath> excludedDirectories, IProgressMonitor monitor) {
+		if (!project.getFile("build.gradle").exists()) return;
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: A comment on style: I should probably clean this line up, and other lines that try to avoid using braces.

**Comment by fbricon** [comment]
The first thing you should do is squash all these commits into one

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java (hunk 1 above)
Comment (fbricon) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java]: Formatting changes make reviews _very_ difficult. Github eases the pain when you append ?w=1 to the PR url, so the diff ignores whitespace changes, see https://github.com/Ian-Stewart-Binks/buildship/pull/1/files?w=1 for instance, but generally speaking you should really really avoid that kind of change.

It seems the OOMPH setup installs some JDT formatter preferences, so try Ctrl+F, see if it  yields better results than CTRL+A / CTRL+I

File: org.eclipse.buildship.ui.importer/build.properties [hunk 8]
```diff
@@ -0,0 +1,9 @@
+source.. = src/main/java/,\
+           src/main/resources/
+output.. = bin/
+bin.includes = .,\
+               help/,\
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/build.properties]: help and icons should be deleted here

File: org.eclipse.buildship.ui.importer/help/help-contexts.xml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<contexts>
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/help/help-contexts.xml]: file to remove

File: org.eclipse.buildship.ui.importer/help/html/project-import.html [hunk 10]
```diff
@@ -0,0 +1,7 @@
+<h1>Importing a Gradle Project</h1>
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/help/html/project-import.html]: file to remove

File: org.eclipse.buildship.ui.importer/plugin.xml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?eclipse version="3.4"?>
+<plugin>
+   <extension
+         point="org.eclipse.e4.ui.importer.projectConfigurators">
+      <projectConfigurator
+            class="org.eclipse.buildship.ui.importer.internal.GradleProjectConfigurator"
+            label="-Gradle-">
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/plugin.xml]: label should most likely be configured in plugin.properties (compare with other buildship plugin.xml labels)

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
+package org.eclipse.buildship.ui.importer.internal;
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Copyright header is missing, check other buildship files for reference

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 7 above)
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Yes please always use curly braces

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 6 above)
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: the nature should be added by the import job, not here

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+	/**
+	 * Sets up the Gradle configurations for the given project.
+	 * @param project
+	 */
+	private ProjectImportConfiguration createGradleProjectConfigurations(IProject project) {
+		ProjectImportConfiguration configuration = new ProjectImportConfiguration();
+		GradleDistributionWrapper gradleDistribution = GradleDistributionWrapper.from(DistributionType.WRAPPER, null);
+
+		configuration.setGradleDistribution(gradleDistribution);
+		configuration.setGradleUserHome(null);
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: this is unnecessary

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+		configuration.setApplyWorkingSets(true);
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: I believe you should refrain from setting default (null or empty) values. I expect ProjectImportConfiguration default state will work OOTB. Only change what needs to be changed.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+		try {
+			projectImportJob.join();
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+	}
+
+	@Override
+	public boolean shouldBeAnEclipseProject(IContainer container, IProgressMonitor monitor) {
+		IFile gradleFile = container.getFile(new Path("build.gradle"));
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Let's start by constraining easymport to projects containing a gradle wrapper

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+		return gradleFile.exists();
+	}
+
+	@Override
+	public Set<IFolder> getDirectoriesToIgnore(IProject project, IProgressMonitor monitor) {
+		Set<IFolder> res = new HashSet<IFolder>();
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: you know the size of the Set, so you can directly do `new HashSet<IFolder>(3)`

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+	public void stop(BundleContext context) throws Exception {
+		plugin = null;
+		unregisterServices();
+		super.stop(context);
+	}
+
+	private void registerServices(BundleContext context) {
+		// store services with low ranking such that they can be overridden
+		// during testing or the like
+		Dictionary<String, Object> preferences = new Hashtable<String, Object>();
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java]: new Hashtable<String, Object>(1)

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+		preferences.put(Constants.SERVICE_RANKING, 1);
+
+		Dictionary<String, Object> priorityPreferences = new Hashtable<String, Object>();
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java]: new Hashtable<String, Object>(1)

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/console/ConsoleMessages.properties [hunk 19]
```diff
@@ -0,0 +1,16 @@
+#
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/console/ConsoleMessages.properties]: file to remove

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/launch/LaunchMessages.properties [hunk 20]
```diff
@@ -0,0 +1,26 @@
+#
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/launch/LaunchMessages.properties]: file to remove

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/projectimport/ProjectImportMessages.properties [hunk 21]
```diff
@@ -0,0 +1,40 @@
+#
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/projectimport/ProjectImportMessages.properties]: file to remove

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/taskview/TaskViewMessages.properties [hunk 22]
```diff
@@ -0,0 +1,38 @@
+#
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/taskview/TaskViewMessages.properties]: file to remove

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/view/execution/ExecutionsViewMessages.properties [hunk 23]
```diff
@@ -0,0 +1,17 @@
+#
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/view/execution/ExecutionsViewMessages.properties]: file to remove

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/viewer/ViewerMessages.properties [hunk 24]
```diff
@@ -0,0 +1,22 @@
+###############################################################################
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/viewer/ViewerMessages.properties]: file to remove

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy [hunk 25]
```diff
@@ -0,0 +1,151 @@
+package org.eclipse.buildship.ui.importer
```
Comment (fbricon) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy]: Copyright header is missing, check other buildship files for reference

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java (hunk 1 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/projectimport/ProjectImportJob.java]: I had no idea about the ?w=1. That's pretty cool.

Ctrl + SHIFT + F yields... different... results. It ensures that lines don't exceed ~80 characters, which is something that the Buildship code does a lot, so it ends up reformatting the code even more. It looks like the Buildship code sticks with spaces, I'll just do that manually and see if I can modify my Eclipse setup to account for that.

File: org.eclipse.buildship.ui.importer/help/help-contexts.xml (hunk 9 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/help/help-contexts.xml]: Removed.

File: org.eclipse.buildship.ui.importer/build.properties (hunk 8 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/build.properties]: Deleted.

File: org.eclipse.buildship.ui.importer/help/html/project-import.html (hunk 10 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/help/html/project-import.html]: Removed.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/console/ConsoleMessages.properties (hunk 19 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/console/ConsoleMessages.properties]: Check.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/launch/LaunchMessages.properties (hunk 20 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/launch/LaunchMessages.properties]: Check.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/projectimport/ProjectImportMessages.properties (hunk 21 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/projectimport/ProjectImportMessages.properties]: Check.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/taskview/TaskViewMessages.properties (hunk 22 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/taskview/TaskViewMessages.properties]: Check.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/view/execution/ExecutionsViewMessages.properties (hunk 23 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/view/execution/ExecutionsViewMessages.properties]: Check.

File: org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/viewer/ViewerMessages.properties (hunk 24 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/resources/org/eclipse/buildship/ui/viewer/ViewerMessages.properties]: Done.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java (hunk 18 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java]: Done.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java (hunk 17 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java]: Done.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 16 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: I didn't know that this was possible. Added it in. Thanks!

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 6 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Hmm. Perhaps there might be a way for the WorkspaceOperations to be able to detect the project description that is supposedly created by the Easymport code. I'll look into this.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 14 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Will do.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 15 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Done, but will need to modify the tests to account for this.

File: org.eclipse.buildship.ui.importer/META-INF/MANIFEST.MF [hunk 26]
```diff
@@ -0,0 +1,32 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Buildship, Eclipse Plug-ins for Gradle - UI Importer
+Bundle-SymbolicName: org.eclipse.buildship.ui.importer;singleton:=true
+Bundle-Version: 1.0.0.qualifier
+Bundle-Vendor: Eclipse Buildship
+Bundle-RequiredExecutionEnvironment: JavaSE-1.6
+Bundle-Activator: org.eclipse.buildship.ui.importer.internal.UiImporterPlugin
+Require-Bundle: org.eclipse.buildship.core,
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/META-INF/MANIFEST.MF]: lots of these requirements should be pruned. Only declare what is really needed

File: org.eclipse.buildship.ui.importer/about.html (hunk 2 above)
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/about.html]: about.html is a mandatory file @ eclipse.org

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+public class GradleProjectConfigurator implements ProjectConfigurator {
+
+	@Override
+	public boolean canConfigure(IProject project, Set<IPath> ignoredPaths, IProgressMonitor monitor) {
+		return shouldBeAnEclipseProject(project, monitor);
+	}
+
+	@Override
+	public void configure(IProject project, Set<IPath> excludedDirectories, IProgressMonitor monitor) {
+		if (!project.getFile("build.gradle").exists()) {
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: you should call shouldBeAnEclipseProject instead

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+	/**
+	 * Sets up the Gradle configurations for the given project.
+	 * @param project
+	 */
+	private ProjectImportConfiguration createGradleProjectConfigurations(IProject project) {
+		ProjectImportConfiguration configuration = new ProjectImportConfiguration();
+		GradleDistributionWrapper gradleDistribution = GradleDistributionWrapper.from(DistributionType.WRAPPER, null);
+
+		configuration.setGradleDistribution(gradleDistribution);
+		configuration.setApplyWorkingSets(true);
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: we're not setting workingsets anywhere during smart import, so you should keep the default behavior

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+		configuration.setWorkingSets(new ArrayList<String>());
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: should be unnecessary

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+	 * @param project
+	 * @param configuration
+	 */
+	private void importProject(IProject project, ProjectImportConfiguration configuration) {
+		ProjectImportJob projectImportJob = new ProjectImportJob(project, configuration);
+		projectImportJob.schedule();
+		
+		try {
+			projectImportJob.join();
+		} catch (InterruptedException e) {
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: no printstacktraces please. Rather use UiImporterPlugin.logger()... instead

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+	 * Imports the given project according to the configuration.
+	 * @param project
+	 * @param configuration
+	 */
+	private void importProject(IProject project, ProjectImportConfiguration configuration) {
+		ProjectImportJob projectImportJob = new ProjectImportJob(project, configuration);
+		projectImportJob.schedule();
+		
+		try {
+			projectImportJob.join();
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: I'm really not too fond of having to join() here. But I believe the smart import needs configure(...) to guarantee configuration is complete before returning. I would prefer if we could double-check that assessment. 
@mickaelistria can probably give some advice here.

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	@Override
+	public boolean shouldBeAnEclipseProject(IContainer container, IProgressMonitor monitor) {
+		IFile gradleBuildFile = container.getFile(new Path("build.gradle"));
+		IFile gradleShellScript = container.getFile(new Path("gradlew"));
+		IFile gradleBatchFile = container.getFile(new Path("gradle.bat"));
+
+		return gradleBuildFile.exists() &&
+			   gradleShellScript.exists() &&
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: gradleShellScript.exists() || gradleBatchFile.exists();

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java [hunk 33]
```diff
@@ -0,0 +1,143 @@
...
+		return res;
+	}
+
+	@Override
+	public IWizard getConfigurationWizard() {
+		return null;
+	}
+
+	// TODO: Remove function, allocate nature adding responsibility to project import.
+	private void addNature(IProject project, String natureId, IProgressMonitor monitor) {
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: unused code should be deleted

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java [hunk 34]
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2015 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Etienne Studer & Donát Csikós (Gradle Inc.) - initial API and implementation and initial documentation
```
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/UiImporterPlugin.java]: add yourself as contributor for the Smart Import feature

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy [hunk 35]
```diff
@@ -0,0 +1,162 @@
...
+	 */
+	IProject createTestProject(projectName, isGradleFile) {
+		def workspace = ResourcesPlugin.getWorkspace()
+		def projectDescription = workspace.newProjectDescription(projectName)
+		def project = workspace.root.getProject(projectName)
+
+		project.create(projectDescription, monitor)
+		project.open(IResource.NONE, monitor)
+
+		if (isGradleFile) addGradleBuildFile(project)
```
Comment (fbricon) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/importer/ImporterTest.groovy]: add wrapper files too

File: tooling-e45.target (hunk 4 above)
Comment (fbricon) [inline tooling-e45.target]: dunno, prolly PDE automatically incrementing that

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 31 above)
Comment (mickaelistria) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: at that point, it doesn't seem related to smart import. The ProjectImportJob is not part of smart import.
However, if you run job.schedule() then job.join() immediately, you're actually doing a synchronous call, so you'd probably rather use job.run()

**Comment by mickaelistria** [comment]
@Ian-Stewart-Binks can you please add an entry in this table to mention your WIP? https://wiki.eclipse.org/E4/UI/Smart_Import#Get_it

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 31 above)
Comment (fbricon) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: this part is called in smart import's  configure(...)  method. Is it allowed to be long running? can configure just schedule a background job for the actual import and return early?

Comment (mickaelistria) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: ah ok.
So, yes, it is allowed to be long running. The interface passes a progressmonitor to most methods. However, there is currently no way to return early, since the configuration may impact the continuation of import/detection.
A compromise is to run synchronously the "structuring" changes (setting the natures, the source and output folders), and to schedule a job to run asynchronously the part that can be lazy (dependency resolution).

**Comment by Ian-Stewart-Binks** [comment]
@mickaelistria I've added another row to the 'Get It' table. Let me know if I've done anything against the rules/formatting. I will make sure to update it once further progress has been made here.

(WIP means work in progress, correct?)

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 31 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: `run(...)` has a 'Discouraged Access' warning, and is apparently not part of the public API. I've been warned that doing this is a dangerous line to cross. If we plan to go the route of synchronously calling ProjectImportJob, should I use `run` or `schedule` + `join`?

As for compromising, it looks like it might take a bit more thought to separate the structuring changes and dependency resolution process. I'm uncertain whether or not this is the correct route to take.

**Comment by mickaelistria** [comment]
Thanks.
Yes WIP is "work in progress".

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 28 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: ProjectImportJob throws an NPE if this isn't set. This line to be exact: 
```
this.workingSets = configuration.getApplyWorkingSets().getValue() ? ImmutableList.copyOf(configuration.getWorkingSets().getValue()) : ImmutableList.<String>of();
```

File: org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java (hunk 29 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/src/main/java/org/eclipse/buildship/ui/importer/internal/GradleProjectConfigurator.java]: Again, NPE thrown if both of these lines aren't present.

File: org.eclipse.buildship.ui.importer/plugin.xml (hunk 11 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/plugin.xml]: I'm not entirely certain what you mean- what is plugin.properties?

Comment (fbricon) [inline org.eclipse.buildship.ui.importer/plugin.xml]: This is used for localization. Take https://github.com/eclipse/m2e-core/blob/master/org.eclipse.m2e.core.ui/plugin.properties for instance. You need to add `Bundle-Localization: plugin` in your manifest, as in https://github.com/eclipse/m2e-core/blob/master/org.eclipse.m2e.core.ui/META-INF/MANIFEST.MF#L8 (and don't forget to add plugin.properties to your build.properties) and then you can use property keys like in https://github.com/eclipse/m2e-core/blob/master/org.eclipse.m2e.core.ui/plugin.xml#L40

Comment (fbricon) [inline org.eclipse.buildship.ui.importer/plugin.xml]: Read https://www.eclipse.org/articles/Article-Internationalization/how2I18n.html

Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.ui.importer/plugin.xml]: Added. Thanks for the article.
